src/apps/bridge/reportBuilder.cpp

```diff
@@ -0,0 +1,497 @@
+#include "task_priorities.h"
+#include "timer_callback_handler.h"
+#include "timers.h"
+#include <string.h>
+#include "aanderaaController.h"
+#include "app_config.h"
+#include "app_pub_sub.h"
+#include "bm_serial.h"
+#include "cbor_sensor_report_encoder.h"
+#include "device_info.h"
+#include "reportBuilder.h"
+
+#define REPORT_BUILDER_QUEUE_SIZE (16)
+#define TOPO_TIMEOUT_MS (1000)
+
+// This is the current max length the cbor buffer can be for the sensor report
+// The limiting factor is the max size we can send via Iridium minus the size
+// of the messages header
+#define MAX_SENSOR_REPORT_CBOR_LEN 315
+
+typedef struct report_builder_element_s {
+  uint64_t node_id;
+  uint8_t sensor_type;
+  void *sensor_data;
+  report_builder_element_s *next;
+  report_builder_element_s *prev;
+} report_builder_element_t;
+
+static aanderaa_aggregations_t NAN_AGG = {.abs_speed_mean_cm_s = NAN,
+                                           .abs_speed_std_cm_s = NAN,
+                                           .direction_circ_mean_rad = NAN,
+                                           .direction_circ_std_rad = NAN,
+                                           .temp_mean_deg_c = NAN};
+
+class ReportBuilderLinkedList {
+  private:
+
+    report_builder_element_t* newElement(uint64_t node_id, uint8_t sensor_type, void *sensor_data, uint32_t sensor_data_size, uint32_t samples_per_report, uint32_t sample_counter);
+    void freeElement(report_builder_element_s **element_pointer);
+    void append(report_builder_element_t *curr, uint64_t node_id, uint8_t sensor_type, void *sensor_data, uint32_t sensor_data_size, uint32_t samples_per_report, uint32_t sample_counter);
+    void addSensorDataToElement(report_builder_element_t *element, uint8_t sensor_type, void *sensor_data, uint32_t sample_counter);
+
+    report_builder_element_t *head;
+    size_t size;
+
+  public:
+
+    ReportBuilderLinkedList();
+    ~ReportBuilderLinkedList();
+
+    bool removeElement(report_builder_element_t *element);
+    bool findElementAndAddSampleToElement(uint64_t node_id, uint8_t sensor_type, void *sensor_data, uint32_t sensor_data_size, uint32_t samples_per_report, uint32_t sample_counter);
+    report_builder_element_t* findElement(uint64_t node_id);
+    void clear();
+};
+
+typedef struct ReportBuilderContext_s {
+  uint32_t _sample_counter;
+  uint32_t _samplesPerReport;
+  uint32_t _transmitAggregations;
+  ReportBuilderLinkedList _reportBuilderLinkedList;
+  uint64_t _report_period_node_list[TOPOLOGY_SAMPLER_MAX_NODE_LIST_SIZE];
+  uint32_t _report_period_num_nodes;
+  uint32_t _report_period_max_network_crc32;
+} ReportBuilderContext_t;
+
+static ReportBuilderContext_t _ctx;
+static QueueHandle_t _report_builder_queue = NULL;
+
+static void report_builder_task(void *parameters);
+
+ReportBuilderLinkedList::ReportBuilderLinkedList() {
+  head = NULL;
+  size = 0;
+}
+
+ReportBuilderLinkedList::~ReportBuilderLinkedList() {
+  report_builder_element_t *element = head;
```

Comment:
Nit: could we just call `ReportBuilderLinkedList::clear()` here?

Reply:
yup!